src/lib/pickles/wrap.ml

```diff
@@ -283,98 +283,352 @@ let deferred_values (type n) ~(sgs : (Backend.Tick.Curve.Affine.t, n) Vector.t)
+             (module Impl)
+             ~assert_16_bits:(Step_verifier.assert_n_bits ~n:16) )
       in
-      Impls.Step.(As_prover.(fun () -> read Boolean.typ res)) )
-  |> Or_error.ok_exn
+
+      (* Use deferred_values_typ and the constant helper to prepare deferred_values
+         for use in the circuit.  We change some [Opt.t] to [Option.t] because that is
+         what Type.t is configured to accept. *)
+      let deferred_values =
+        constant deferred_values_typ
+          { deferred_values with
+            plonk =
+              { deferred_values.plonk with
+                lookup = Opt.to_option deferred_values.plonk.lookup
+              ; optional_column_scalars =
+                  Composition_types.Wrap.Proof_state.Deferred_values.Plonk
+                  .In_circuit
+                  .Optional_column_scalars
+                  .map ~f:Opt.to_option
+                    deferred_values.plonk.optional_column_scalars
+              }
+          }
+      (* Prepare all of the evaluations (i.e. all of the columns in the proof that we open)
+         for use in the circuit *)
+      and evals =
+        constant
+          (Plonk_types.All_evals.typ (module Impls.Step) feature_flags)
+          { evals = { public_input = x_hat_evals; evals = proof.openings.evals }
+          ; ft_eval1 = proof.openings.ft_eval1
+          }
+      in
+
+      (* Run the circuit without generating a proof using run_and_check *)
+      Impls.Step.run_and_check (fun () ->
+          (* Set up the step sponge from the wrap sponge -- we cannot use the same poseidon
+             sponge in both step and wrap because they have different fields.
+
+             In order to continue the Fiat-Shamir heuristic across field boundaries we use
+             the wrap sponge for everything in the wrap proof, squeeze it one final time and
+             expose the squoze value in the public input to the step proof, which absorbs
+             said squoze value into the step sponge. :-) This means the step sponge has absorbed
+             everything from the proof so far by proxy and that is also over the native field! *)
+          let res, _chals =
+            let sponge =
+              let open Step_main_inputs in
+              let sponge = Sponge.create sponge_params in
+              Sponge.absorb sponge
+                (`Field (Impl.Field.constant sponge_digest_before_evaluations)) ;
+              sponge
+            in
+
+            (* Call finalisation with all of the required details *)
+            Step_verifier.finalize_other_proof
+              (module Nat.N0)
+              ~feature_flags
+              ~step_domains:
+                (`Known
+                  [ { h = Pow_2_roots_of_unity vk.domain.log_size_of_group } ]
+                  )
+              ~sponge ~prev_challenges:[] deferred_values evals
+          in
+
+          (* Actually do the above as the prover *)
+          Impls.Step.(As_prover.(fun () -> read Boolean.typ res)) )
```

Comment:
Nit: `read the result boolean from the circuit and make it available to the OCaml world`

Reply:
Isn't that done by:

Reply:
Got it 🙏